### DISCUSSION TITLE
Run A2DP on ESP32-S3?

### DISCUSSION THREAD

**Discussion by mcbpalisade** [pschatzmann/ESP32-A2DP#566] [Q&A] [answered]
I understand that this chip has dual mode Bluetooth, that is classic and BLE.  Classic is the only one you can use for audio.  Has anyone got audio source transmit to a sink working on this chip?  Thanks.

**Comment by pschatzmann** [comment, marked as answer]
https://github.com/pschatzmann/ESP32-A2DP/wiki/Support-for-esp32-c3%3F

You are mistaken: just read the data sheet!

**Reply by mcbpalisade** [reply]
Thank you, an LLM lied to me and said it did.

So I'll assume that I'm using the only choice the Adafruit WROOM ESP32E feather board.

**Reply by mcbpalisade** [reply]
Just have to ask: Is anyone else using another chip for a solution?  I find it hard to believe that all modern BT source/sink audio products use this [old & relatively expensive] chip.

**Reply by pschatzmann** [reply]
I don't understand your question: on AliExpress you can find ESP32 Dev Modules starting from 2 USD and I don't think this is expensive for what you get...

**Reply by mcbpalisade** [reply]
My question is: Is there an inclusive list for the chips that can run A2DP?  I'm really only interested in the Arduino types, not so much the Nordic family et al.  I really don't care what they cost but I avoid all the Alibaba store fronts, generally.

I'm using the Adafruit 3405 product now to which I add an Adalogger which has an SD card slot.  It costs $20 for the processor which is a lot compared to your $2 suggestion.  I'm doing avionics development so cost isn't the first consideration.

**Comment by frankcohen** [comment]
Thank you for creating and maintaining ESP32-A2DP. I appreciate your effort. I use it in my [AirGuitar project](https://github.com/frankcohen/ReflectionsOS/tree/main/Experiments/AirGuitar). I agree the ESP32 datasheet is the final word on A2DP support in non-classic ESP32 BLE, generative AI is giving bad answers. For example, I ask Github Copilot:

`I need a Custom BLE Audio Implementation for ESP32-S3 that does A2DP and compiles in Arduino IDE 2 and does not use pschatzmann library
`

and Copilot responds:


`To create a custom BLE audio implementation for the ESP32-S3 that uses A2DP and compiles in the Arduino IDE 2 without using the pschatzmann library, we will need to directly utilize the Bluetooth stack provided by Espressif in the Arduino framework. Below is a detailed example that sets up an A2DP source to send audio data to a Bluetooth speaker.

Arduino Sketch
Create a new sketch in the Arduino IDE and name it A2DP_Source.ino.


A2DP_Source.ino
v4
#include <Arduino.h>
#include "BluetoothSerial.h"
#include "esp_bt_main.h"
#include "esp_bt_device.h"
#include "esp_a2dp_api.h"
Explanation
Initialization:

The NVS (Non-Volatile Storage) is initialized.
The Bluetooth stack is initialized and enabled.
A2DP source callbacks are registered.
I2S Configuration:

The I2S peripheral is configured to transmit audio data. Adjust the pin numbers and settings as needed for your specific hardware setup.
Sending Audio Data:

In the loop function, audio data is simulated by filling the audio_data buffer with zeros. Replace this with actual audio data for your application.
The bt_app_a2d_data_cb function sends the audio data to the connected Bluetooth speaker.
Upload the Sketch
Connect your ESP32-S3 board to your computer.
Select the correct board and port under Tools > Board and Tools > Port.
Click the upload button to upload the sketch to your ESP32-S3 board.
Once uploaded, your ESP32-S3 should start broadcasting as an A2DP source. You can then connect a Bluetooth speaker to the ESP32-S3 and start streaming the audio data.

Feel free to replace the placeholder audio data with actual audio data to test the A2DP functionality. If you have any further questions or need additional features, please let me know!
`

So helpful, and wrong.

-Frank

**Reply by pschatzmann** [reply]
It is a well known fact that all AI models are just hallucinating and come up with some wrong random answers.

Don't trust anything w/o double checking !

**Reply by frankcohen** [reply]
> It is a well known fact that all AI models are just hallucinating and come up with some wrong random answers.
> 
> Don't trust anything w/o double checking !

Thanks! ESP32-H4 appears to solve the BLE Audio issue. It provides Bluetooth BLE 5.2, including audio and fun things you can do with direction finding. Espressif told me to expect it in the second quarter of 2025. Here's the announcement. https://www.espressif.com/en/news/ESP32-H4
